include/emds/emds.h

```diff
@@ -0,0 +1,183 @@
+		.id = _id,                                                                         \
+		.data = (uint8_t *)_data,                                                          \
+		.len = _len,                                                                       \
+	}
+
+/**
+ * @brief emds_init Initialize the emergency data storage.
+ *
+ * Initializes the emergency data storage. This need to called before starting
+ * to add entries to the emds_dynamic_entry and doing restore.
+ *
+ * @retval 0 Success
+ * @retval -ERRNO errno code if error
+ */
+int emds_init(void);
+
+/**
+ *  @brief Add entry that should be saved/restored when emergency data storage is called.
+ *
+ * This will add the entry to the dynamic entry list, and will when the
+ * 'emds_store' function is called take the given data pointer and store the
+ * data to the emergency data storage.
+ *
+ * @param entry Entry to add to list.
+ */
+void emds_entry_add(struct emds_dynamic_entry *entry);
+
+/**
+ * @brief Start the emergency data storage process.
+ * 
+ * This will trigger the process of storing all data that has been regitster to
+ * be stored. When it is finished with storing the data the data should not be
+ * changed and the board should be halted or rebooted at the end. This is a time
+ * critical process and will be processed as fast as possible. This will store
+ * all data register either though the 'emds_entry_add' or 'EMDS_ENTRY_DEFINE'.
+ */
+void emds_store(void);
+
+/**
+ * @brief Load all data from the emergency data storage.
+ *
+ * This need to be called after the static and dynamic entries are added, as
+ * they are used to select data that should be loaded. This should also be
+ * called before the 'emds_prepare' function is called as that will delete all 
+ * the stored data.
+ *
+ * @retval 0 Success
+ * @retval -ERRNO errno code if error
+ */
+int emds_load(void);
+
+/**
+ * @brief Clear flash area from the emergency data storage.
+ *
+ * This function will clear the flash area for all previous stored data.
+ *
+ * @retval 0 Success
+ * @retval -ERRNO errno code if error
+ */
+int emds_clear(void);
+
+/**
+ * @brief Prepare flash area for a next emergency data storage.
+ *
+ * This function will prepare the flash area for the next emergency data
+ * storage. This delete the current entries if it enough space for the next
+ * emergency data storage. It will clear the flash storage if it is not enough
+ * space for the next storage. This has to be done after all the dynamic is
+ * added.
+ *
+ * @retval 0 Success
+ * @retval -ERRNO errno code if error
+ */
+int emds_prepare(void);
+
+/**
+ * @brief Calculate the time needed to store the registered data.
+ *
+ * This will calculate how much time it takes to store all registered data, both
+ * dynamic and static data registered in the entries.
+ *
+ * @return Time in micro seconds that is needed to store all data.
+ */
+int emds_store_time_get(void);
```

Comment:
if time is in microseconds it is better to use a wider type. uint64_t probably?

Reply:
It should be enough with a uint32_t, can not see that it is necessary with a uint64_t. This should not take more than 4000 seconds.